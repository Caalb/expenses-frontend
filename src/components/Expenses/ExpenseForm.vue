<template>
  <Form
    @submit="onSubmit"
    class="q-gutter-md"
    :validation-schema="validationSchema"
    :initial-values="formattedValues"
    for="expense-form"
    data-testid="expense-form"
  >
    <Field
      name="description"
      v-slot="{ field, errorMessage , value}"
    >
      <QInput
        outlined
        v-bind="field"
        :model-value="value"
        label="Descrição"
        :error-message="errorMessage"
        :error="!!errorMessage"
        data-testid="description-input"
        for="description-input"
      />
    </Field>

    <Field
      name="date"
      v-slot="{ field, errorMessage, value }"
    >
      <QInput
        outlined
        v-bind="field"
        label="Data"
        type="date"
        :model-value="value"
        :error-message="errorMessage"
        :error="!!errorMessage"
        :max="new Date().toISOString().split('T')[0]"
        data-testid="date-input"
        for="date-input"
      />
    </Field>

    <Field
      name="amount"
      v-slot="{ field, errorMessage, value }"
    >
      <QInput
        outlined
        v-bind="field"
        label="Valor"
        :model-value="value"
        type="number"
        step="0.01"
        prefix="R$"
        :error-message="errorMessage"
        :error="!!errorMessage"
        data-testid="amount-input"
        for="amount-input"
      />
    </Field>

    <div class="row justify-end q-gutter-sm">
      <QBtn
        label="Cancelar"
        color="grey-6"
        @click="$emit('cancel')"
        data-testid="cancel-button"
      />
      <QBtn
        type="submit"
        :label="isEditing ? 'Salvar' : 'Criar'"
        color="primary"
        :loading="submitting"
        data-testid="submit-button"
      />
    </div>
  </Form>
</template>

<script setup lang="ts">
import { QInput, QBtn } from 'quasar'
import { Form, Field } from 'vee-validate'
import { toTypedSchema } from '@vee-validate/zod'
import { z } from 'zod'
import { formatDate } from '@/helpers/formatDate'
import type { ExpenseFormData } from '@/types/expense'

const props = defineProps<{
  initialValues: ExpenseFormData
  isEditing: boolean
  submitting: boolean
}>()

const formattedValues = {
  date: props.initialValues.date ? formatDate(props.initialValues.date, "yyyy-MM-dd") : "",
  amount: (props.initialValues.amount as number | string) === 'number'
    ? (props.initialValues.amount as unknown as number).toString()
    : (props.initialValues.amount as string) || "",
  description: props.initialValues.description || ""
}
const emit = defineEmits<{
  (e: 'submit', values: ExpenseFormData): void
  (e: 'cancel'): void
}>()

const validationSchema = toTypedSchema(
  z.object({
    description: z.string()
      .min(3, 'Descrição deve ter no mínimo 3 caracteres')
      .max(191, 'Descrição deve ter no máximo 191 caracteres'),

    date: z.string().min(1, 'Data é obrigatória'),

    amount: z.preprocess(
      (val) => (typeof val === 'string' ? val : String(val)),
      z.string()
        .min(1, 'Valor é obrigatório')
        .refine((val) => !isNaN(Number(val)), 'Valor deve ser um número válido')
        .refine((val) => Number(val) > 0, 'Valor deve ser maior que zero')
        .transform((val) => Number(parseFloat(val).toFixed(2)))
    )
  })
)

const onSubmit = (values: unknown) => {
  const expenseValues = values as ExpenseFormData
  emit('submit', expenseValues)
}
</script>
