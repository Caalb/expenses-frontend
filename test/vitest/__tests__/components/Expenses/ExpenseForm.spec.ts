import { mount, VueWrapper } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { getByTestId } from 'test/vitest/__utils__'
import { describe, expect, it, beforeEach, vi } from 'vitest'
import ExpenseForm from '@/components/Expenses/ExpenseForm.vue'
import { Form } from 'vee-validate'

installQuasarPlugin()

const createWrapper = () =>
  mount(ExpenseForm, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
    },
    props: {
      initialValues: {
        id: 1,
        amount: '100',
        date: '2023-01-01',
        description: 'Expense 1',
      },
      isEditing: false,
      submitting: false,
    },
  }) as VueWrapper

describe('ExpenseForm', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = createWrapper()
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit cancel when cancel button is clicked', async () => {
    await wrapper.vm.$nextTick()

    const cancelButton = getByTestId(wrapper, 'cancel-button')
    expect(cancelButton.exists()).toBe(true)

    await cancelButton.trigger('click')

    expect(wrapper.emitted('cancel')).toBeTruthy()
  })

  it('should emit submit when submit button is clicked and valid form', async () => {
    const submitButton = getByTestId(wrapper, 'submit-button')
    expect(submitButton.exists()).toBe(true)

    await getByTestId(wrapper, 'amount-input').setValue('100')
    await getByTestId(wrapper, 'date-input').setValue('2023-01-01')
    await getByTestId(wrapper, 'description-input').setValue('Expense 1')

    wrapper.findComponent(Form).vm.$emit('submit', {
      amount: '100',
      date: '2023-01-01',
      description: 'Expense 1',
    })

    expect(wrapper.emitted('submit')).toBeTruthy()
    expect(wrapper.emitted('submit')?.[0]).toEqual([
      {
        amount: '100',
        date: '2023-01-01',
        description: 'Expense 1',
      },
    ])
  })
})
