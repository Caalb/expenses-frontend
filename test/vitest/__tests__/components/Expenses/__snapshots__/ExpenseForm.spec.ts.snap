// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpenseForm > should match snapshot 1`] = `
"<form novalidate="" class="q-gutter-md" for="expense-form" data-testid="expense-form"><label class="q-field row no-wrap items-start q-field--outlined q-input q-field--float q-field--labeled q-field--with-bottom" for="description-input">
    <!---->
    <div class="q-field__inner relative-position col self-stretch">
      <div class="q-field__control relative-position row no-wrap" tabindex="-1">
        <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Descrição" name="description" data-testid="description-input" id="description-input" type="text" value="Expense 1">
          <div class="q-field__label no-pointer-events absolute ellipsis">Descrição</div>
          <!---->
        </div>
      </div>
      <div class="q-field__bottom row items-start q-field__bottom--animated">
        <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
          <div class="q-field__messages col"></div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
  </label><label class="q-field row no-wrap items-start q-field--outlined q-input q-field--float q-field--labeled q-field--with-bottom" for="date-input">
    <!---->
    <div class="q-field__inner relative-position col self-stretch">
      <div class="q-field__control relative-position row no-wrap" tabindex="-1">
        <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Data" name="date" max="2024-11-07" data-testid="date-input" id="date-input" type="date" value="2022-12-31">
          <div class="q-field__label no-pointer-events absolute ellipsis">Data</div>
          <!---->
        </div>
      </div>
      <div class="q-field__bottom row items-start q-field__bottom--animated">
        <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
          <div class="q-field__messages col"></div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
  </label><label class="q-field row no-wrap items-start q-field--outlined q-input q-field--float q-field--labeled q-field--with-bottom" for="amount-input">
    <!---->
    <div class="q-field__inner relative-position col self-stretch">
      <div class="q-field__control relative-position row no-wrap" tabindex="-1">
        <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
          <div class="q-field__prefix no-pointer-events row items-center">R$</div><input class="q-field__native q-placeholder" tabindex="0" aria-label="Valor" name="amount" step="0.01" data-testid="amount-input" id="amount-input" type="number" value="100">
          <div class="q-field__label no-pointer-events absolute ellipsis">Valor</div>
          <!---->
        </div>
      </div>
      <div class="q-field__bottom row items-start q-field__bottom--animated">
        <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
          <div class="q-field__messages col"></div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
  </label>
  <div class="row justify-end q-gutter-sm"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-grey-6 text-white q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button" data-testid="cancel-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Cancelar</span></span></button><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable" tabindex="0" type="submit" data-testid="submit-button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Criar</span></span>
      <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
        <!---->
      </transition-stub>
    </button></div>
</form>"
`;
